### ISSUE TITLE
No ability for clients to request program heap sizes

### ISSUE THREAD

**Issue by jackcmay** [solana-labs/solana#20298]
#### Problem

Some programs may require larger heaps than the default 32k.  One example is the neon VM which uses a special case (#17700) that should be replaced with something universal.

#### Proposed Solution

The compute budget program supports requesting a tx wide bump in max compute units.  A custom heap could also be requested in a similar manner.

Doing so would require a new compute budget instruction something like:

```
/// Request a transaction wide program heap size
RequestHeap(u64),
```

The requested heap would apply transaction-wide, so all the instructions within the transaction would get the larger heap.  The result is programs that don't take advantage of the larger heap would be unaffected, but there would be an added runtime cost associated with allocating the larger amount of heap for each instruction during virtual machine creation.

An alternative approach would be to request a bump for a particular program or programs within the tx:

```
/// Request a program specific heap size
RequestHeap(u64, Pubkey),
```

Multiple of these instructions could be included to request custom heap sizes per program in the instruction.  All other programs would receive the default heap size.

Thoughts?

**Comment by mvines** [comment]
> The requested heap would apply transaction-wide, so all the instructions within the transaction would get the larger heap.

This approach seems simpler to implement, not sure we'd need/want the extra flexibly that the alternative approach provides

**Comment by jackcmay** [comment]
Thanks @mvines for the quick feedback!

I think both are pretty simple, the latter would contain a list of heap/program id pairs so the side-effect would be searching that list for each instruction in the transaction.

**Comment by mvines** [comment]
Perhaps a program could declare its minimum stack size?  If my program requires more than the default, now there's a new attack surface I need to worry about.  My heap (and stack size) could be adjusted by the caller until they potentially find a magic size that causes my program to misbehave in an exploitable way.

**Comment by jackcmay** [comment]
Yeah, that could be an attack surface, not sure how a program could declare its minimum.

**Comment by jackcmay** [comment]
Though I guess the opposite could be true, program invocation that would normally fail due to heap exhaustion may complete if a larger heap is requested

**Comment by jackcmay** [comment]
@jstarry from a malicious perspective, what are your thoughts on allowing tx writers request higher heaps?

**Comment by jstarry** [comment]
> @jstarry from a malicious perspective, what are your thoughts on allowing tx writers request higher heaps?

Assuming it can be requested at any CPI depth, main concern is resource exhaustion I suppose, so we'd need some sort of cap. Other than that, as long as the heap increase doesn't overlap with or modify the input buffers region, then I think it should be ok.

**Comment by jackcmay** [comment]
@jstarry The heap bump would apply transaction-wide, which means for all instructions in the tx and all CPIs those instructions would make.

**Comment by jstarry** [comment]
> @jstarry The heap bump would apply transaction-wide, which means for all instructions in the tx and all CPIs those instructions would make.

Ah, sorry I didn't review your proposal closely enough. Can we make CPI invocation cost proportional to the requested heap increase?

**Comment by jackcmay** [comment]
Yeah, we could do that, there would be an associated cost to the higher heaps across the board
